### PR TITLE
Added support for Rhubarb Lip Sync TSV and XML file formats for ValueNode_AnimatedFile

### LINKED ISSUE #2169: Rhubarb suport 🙃
Hi
Two days ago I made a topic in synfig forms for Margeing papagayo and synfig

● So in my opinion papagayo is not a stable program and it has lots of bugs and problems I it has been a week that iam searching for a way to make papagayo work correctly but there was no result 🥲
But then today I found a program called rhubarb it has alot of features and my only problem with it was the lack of gui but that’s OK it just makes some things a little harder

■ Rhubarb is supported by lot of animation softwares (opentoonz moho adobe Ae) :sunglasses:🤌and it is new and more stable​:muscle: and also it dosent need a duiloge it’s funny that it done the same thing that I did in last topic :grinning_face_with_smiling_eyes:

:heart: My point is if you can make synfig to suport rhubarb exported formats that would be nice (or maybe you can add it as a plug-in or a tool):smiling_face_with_three_hearts:

Thank Rodolforg for suggesting me to post this here 😃

### PR DESCRIPTION
**Pull request by rodolforg**
fix #2169

Rhubarb Lip Sync:
https://github.com/DanielSWolf/rhubarb-lip-sync

### CODE REVIEW

**Comment by rodolforg** [comment]
@b-aaz Do you use Windows or Linux/MacOS?

If you use Windows, could you please test this PR?  Hint: https://synfig.readthedocs.io/en/latest/how_to_help/testing_prs.html
Otherwise, could you please build and test it?

**Comment by b-aaz** [comment]
Hi happy to see the new update I use both Linux and windows 
So I should download this version and check it with rhubarb in windows? Iam really happy to help

**Comment by b-aaz** [comment]
> Hi happy to see the new update I use both Linux and windows 
> So I should download this version and check it with rhubarb in windows? Iam really happy to help 

Iam downloading the file from appveyor

**Comment by b-aaz** [comment]
> @b-aaz Do you use Windows or Linux/MacOS?
> 
> If you use Windows, could you please test this PR?  Hint: https://synfig.readthedocs.io/en/latest/how_to_help/testing_prs.html
> Otherwise, could you please build and test it?

Hi I tested the new build 
First of all the new gui is awesome 
It didn't crash 
I tried both xml and tsv
The details were grate 
But the x layer or the layer that is used when the audio is silent didn't work it's name got bold but nothing appeared 
All the other phonemes worked  

It said I didn't have ffmpeg when I wanted to render but I had ffmeg idont think that’s a important problem 

I was amazed how much easier it became to lipsync thank you 😊🙏❤
I will continue testing but I didn't see any other points till now

**Comment by rodolforg** [comment]
>  But the x layer or the layer that is used when the audio is silent didn't work it's name got bold but nothing appeared

There is no special treatment for any phoneme or X layer. Did you place the image that represents the shut mouth there? XD




> It said I didn't have ffmpeg when I wanted to render but I had ffmeg idont think that’s a important problem

This 'AppVeyor' version is raw ;) That's why it said that.

File: synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp [hunk 1]
```diff
@@ -130,21 +168,138 @@ class ValueNode_AnimatedFile::Parser
...
+	/// \param[in] s the XML stream to be parsed
+	/// \param[out] value the map of time -> layer name
+	/// \param[out] fields the map of metafield -> value
+	/// \return true if successful. There is no guarantee about stream offset in case of failure
+	static bool parse_xml(istream &s, map<Time, String> &values, map<String, String> &fields)
+	{
+		try {
+			xmlpp::DomParser parser;
+			parser.parse_stream(s);
+			if (parser) {
```
Comment (ice0) [inline synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp:235]: I think better to quit early:
```
if (!parser) {
    return false;
}
```

File: synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp [hunk 2]
```diff
@@ -130,21 +168,138 @@ class ValueNode_AnimatedFile::Parser
...
+					synfig::error(_("XML animated file: it only supports Rhubarb Lip Sync files. Found %s"), root->get_name().c_str());
+					return false;
+				}
+				if (auto metadata_node = root->get_first_child("metadata")) {
+					if (auto sound_node = metadata_node->get_first_child("soundFile")) {
+						if (auto sound_element = dynamic_cast<xmlpp::Element*>(sound_node)) {
+							if (auto sound_text_node = sound_element->get_child_text())
+								fields["sound"] = sound_text_node->get_content();
+						}
+						if (fields.count("sound"))
```
Comment (ice0) [inline synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp:254]: It looks like it should warn if sound element doesn't load? 
```
if (!fields.count("sound"))
```

File: synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp [hunk 3]
```diff
@@ -130,21 +168,138 @@ class ValueNode_AnimatedFile::Parser
...
+									synfig::warning(_("XML animated file: missing attribute 'start' for Rhubarb mouthcue element: skipping"));
+								}
+							}
+						}
+					}
+				}
+				return true;
+			}
+		} catch(xmlpp::internal_error &x) {
+			if (std::string(x.what()) == "Couldn't create parsing context")
```
Comment (ice0) [inline synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp:282]: I think it's better to report a error even if it doesn't match the description. I mean:
```
if (std::string(x.what()) == "Couldn't create parsing context")
    synfig::error(_("XML animated file: Can't XML open file"));
else
    synfig::error(x.what());
```

**Review by ice0 — CHANGES_REQUESTED** [review]

File: synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp (hunk 1 above)
Comment (rodolforg) [reply synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp:235]: ok

File: synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp (hunk 2 above)
Comment (rodolforg) [reply synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp:254]: Whoops!

Comment (rodolforg) [reply synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp:254]: I'll fix the commit and then push --force

Comment (rodolforg) [reply synfig-core/src/synfig/valuenodes/valuenode_animatedfile.cpp]: done!

**Comment by rodolforg** [comment]
@ice0  I think now we only need @b-aaz feedback about this:

> > But the x layer or the layer that is used when the audio is silent didn't work it's name got bold but nothing appeared
> 
> There is no special treatment for any phoneme or X layer. Did you place the image that represents the shut mouth there? XD

**Comment by rodolforg** [comment]
What about we let TSV to load animation values for real, integer, boolean, angle and time values?

**Comment by ice0** [comment]
> What about we let TSV to load animation values for real, integer, boolean, angle and time values?

Can you explain this in more detail? 
It looks like [TSV](https://github.com/DanielSWolf/rhubarb-lip-sync#tsv) only needs real (timestamp) and text (for mouth shape)

**Comment by rodolforg** [comment]
I recall somebody asking for "Animation File" converter to any parameter.
https://forums.synfig.org/t/conversion-to-animation-from-file/11364

The idea is not related to "Import" action, but for the already available converter that Import indirectly uses.
User could write a TSV like:
```
0.00	1.0
0.30	1.5
0.80	1.2
```
The he could convert the Radius parameter of a Circle layer, for example, to "Animation from File", and then referring to the above TSV file.

**Comment by b-aaz** [comment]
> >  But the x layer or the layer that is used when the audio is silent didn't work it's name got bold but nothing appeared
> 
> There is no special treatment for any phoneme or X layer. Did you place the image that represents the shut mouth there? XD
> 
> > It said I didn't have ffmpeg when I wanted to render but I had ffmeg idont think that’s a important problem
> 
> This 'AppVeyor' version is raw ;) That's why it said that.

Yeap I tried the example circle and my own moth shape but nothing showed up tried resizing to make sure nothing is out of canvas but everything was in the canvas

**Comment by b-aaz** [comment]
> > >  But the x layer or the layer that is used when the audio is silent didn't work it's name got bold but nothing appeared
> > 
> > There is no special treatment for any phoneme or X layer. Did you place the image that represents the shut mouth there? XD
> > 
> > > It said I didn't have ffmpeg when I wanted to render but I had ffmeg idont think that’s a important problem
> > 
> > This 'AppVeyor' version is raw ;) That's why it said that.
> 
> Yeap I tried the example circle and my own moth shape In X group but nothing showed up tried resizing to make sure nothing is out of canvas but everything was in the canvas 

I tried it for the 3rd time and it worked I don't know what happened at last two times but I think it's working now 🤔

**Comment by b-aaz** [comment]
https://www.mediafire.com/file/szy0srlnp3e2dga/2021-07-07+18-35-57.mkv/file 
My first lipsync

**Comment by rodolforg** [comment]
Last forced-push I just rebased to master.

**Comment by ice0** [comment]
> Then he could convert the Radius parameter of a Circle layer, for example, to "Animation from File", and then referring to the above TSV file. Thus, the circle radius will be animated by the file contents.

This is interesting idea.

**Review by ice0 — APPROVED** [review]

**Comment by ice0** [comment]
Merged. Thank you!

**Pull request merged**